File: 238.除自身以外数组的乘积/solution.cpp
# include "..\leetcode.h"

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        // right[i] nums[i]右边所有数字的乘积
        vector<int> right(n, 1);
        for(int i = n - 2; i >= 0; i--) {
            right[i] = right[i + 1] * nums[i + 1];
        }
        int left = 1;
        for(int i = 0; i < n; i++) {
            right[i] = left * right[i];
            left *= nums[i];
        }
        return right;
    }
};

File: 238.除自身以外数组的乘积/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        right = [1] * n
        for i in range(n - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1]
        left = 1
        for i in range(0, n):
            right[i] = left * right[i]
            left *= nums[i]
        return right